Sakshi Sharnagat(CS23189)

to apply support vector machine for classification and analyze the impact of differntkernels on model accuracy

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn import metrics
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
from sklearn.metrics import roc_curve
from sklearn.metrics import roc_auc_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score


In [ ]:
from google.colab import drive
drive.mount('/content/drive')


In [ ]:
path="/content/drive/MyDrive/ml 06/pulsar_stars.csv"
df=pd.read_csv(path)
df.head(5)

Mean of the integrated profile  \
0                       140.562500   
1                       102.507812   
2                       103.015625   
3                       136.750000   
4                        88.726562   

    Standard deviation of the integrated profile  \
0                                      55.683782   
1                                      58.882430   
2                                      39.341649   
3                                      57.178449   
4                                      40.672225   

    Excess kurtosis of the integrated profile  \
0                                   -0.234571   
1                                    0.465318   
2                                    0.323328   
3                                   -0.068415   
4                                    0.600866   

    Skewness of the integrated profile   Mean of the DM-SNR curve  \
0                            -0.699648                   3.199833   
1                            -0.515088                   1.677258   
2                             1.051164                   3.121237   
3                            -0.636238                   3.642977   
4                             1.123492                   1.178930   

    Standard deviation of the DM-SNR curve  \
0                                19.110426   
1                                14.860146   
2                                21.744669   
3                                20.959280   
4                                11.468720   

    Excess kurtosis of the DM-SNR curve   Skewness of the DM-SNR curve  \
0                              7.975532                      74.242225   
1                             10.576487                     127.393580   
2                              7.735822                      63.171909   
3                              6.896499                      53.593661   
4                             14.269573                     252.567306   

   target_class  
0             0  
1             0  
2             0  
3             0  
4             0

In [ ]:
df.shape

(17898, 9)

In [ ]:
df.columns

Index([' Mean of the integrated profile',
       ' Standard deviation of the integrated profile',
       ' Excess kurtosis of the integrated profile',
       ' Skewness of the integrated profile', ' Mean of the DM-SNR curve',
       ' Standard deviation of the DM-SNR curve',
       ' Excess kurtosis of the DM-SNR curve', ' Skewness of the DM-SNR curve',
       'target_class'],
      dtype='object')

In [ ]:
X = df.drop('target_class', axis=1)
y = df['target_class']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
df.columns = df.columns.str.strip()
df.columns

Index(['Mean of the integrated profile',
       'Standard deviation of the integrated profile',
       'Excess kurtosis of the integrated profile',
       'Skewness of the integrated profile', 'Mean of the DM-SNR curve',
       'Standard deviation of the DM-SNR curve',
       'Excess kurtosis of the DM-SNR curve', 'Skewness of the DM-SNR curve',
       'target_class'],
      dtype='object')

The column names have been cleaned. Now let's separate the features (X) and the target variable (y) and split the data into training and testing sets.

Here's a mapping for the column names to make them shorter:

In [ ]:
new_column_names = {
    'Mean of the integrated profile': 'Mean_IP',
    'Standard deviation of the integrated profile': 'StdDev_IP',
    'Excess kurtosis of the integrated profile': 'EKurt_IP',
    'Skewness of the integrated profile': 'Skew_IP',
    'Mean of the DM-SNR curve': 'Mean_DMSNR',
    'Standard deviation of the DM-SNR curve': 'StdDev_DMSNR',
    'Excess kurtosis of the DM-SNR curve': 'EKurt_DMSNR',
    'Skewness of the DM-SNR curve': 'Skew_DMSNR'
}

df.rename(columns=new_column_names, inplace=True)
X.rename(columns=new_column_names, inplace=True)

print("Updated DataFrame columns:")
display(df.columns)

Updated DataFrame columns:


Index(['Mean_IP', 'StdDev_IP', 'EKurt_IP', 'Skew_IP', 'Mean_DMSNR',
       'StdDev_DMSNR', 'EKurt_DMSNR', 'Skew_DMSNR', 'target_class'],
      dtype='object')

In [ ]:
df.isnull().any().any()

np.False_

In [ ]:
df.isnull().sum()

Mean_IP         0
StdDev_IP       0
EKurt_IP        0
Skew_IP         0
Mean_DMSNR      0
StdDev_DMSNR    0
EKurt_DMSNR     0
Skew_DMSNR      0
target_class    0
dtype: int64

In [ ]:
df[df.isnull().any(axis = 1)]

Empty DataFrame
Columns: [Mean_IP, StdDev_IP, EKurt_IP, Skew_IP, Mean_DMSNR, StdDev_DMSNR, EKurt_DMSNR, Skew_DMSNR, target_class]
Index: []

In [ ]:
X = df.drop(columns=['target_class'])
y = df['target_class']

In [ ]:
print(y.value_counts())
print(y.value_counts(normalize=True) * 100)

target_class
0    16259
1     1639
Name: count, dtype: int64
target_class
0    90.842552
1     9.157448
Name: proportion, dtype: float64


This output shows the count and percentage of each class in the `target_class` variable, indicating whether the dataset is balanced or imbalanced. We can see that the majority class (0) is significantly larger than the minority class (1), indicating an imbalanced dataset.

The final output confirms that the dataset is imbalanced. Class '0' constitutes about 90.84% of the data, while class '1' makes up only about 9.16%.

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("Features scaled successfully!")

Features scaled successfully!


The features have been scaled using `StandardScaler`. Now, we can proceed with training the Support Vector Machine (SVM) models.

In [ ]:
from sklearn.svm import SVC

# Initialize SVC with a linear kernel and C=1.0
svc_linear = SVC(kernel='linear', C=1.0, random_state=42)

# Train the model
svc_linear.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred_linear = svc_linear.predict(X_test_scaled)

# Evaluate the model
accuracy_linear = accuracy_score(y_test, y_pred_linear)
conf_matrix_linear = confusion_matrix(y_test, y_pred_linear)
class_report_linear = classification_report(y_test, y_pred_linear)

print(f"Accuracy (Linear Kernel, C=1.0): {accuracy_linear:.4f}")
print("\nConfusion Matrix (Linear Kernel, C=1.0):\n", conf_matrix_linear)
print("\nClassification Report (Linear Kernel, C=1.0):\n", class_report_linear)

Accuracy (Linear Kernel, C=1.0): 0.9788

Confusion Matrix (Linear Kernel, C=1.0):
 [[3243   16]
 [  60  261]]

Classification Report (Linear Kernel, C=1.0):
               precision    recall  f1-score   support

           0       0.98      1.00      0.99      3259
           1       0.94      0.81      0.87       321

    accuracy                           0.98      3580
   macro avg       0.96      0.90      0.93      3580
weighted avg       0.98      0.98      0.98      3580



In [ ]:
X_train

Mean of the integrated profile  \
14453                      108.765625   
15797                      121.554688   
4510                       100.265625   
11899                      118.460938   
895                         97.109375   
...                               ...   
11284                      124.312500   
11964                      115.617188   
5390                       116.031250   
860                        135.664062   
15795                      120.726562   

       Standard deviation of the integrated profile  \
14453                                     47.520470   
15797                                     50.909357   
4510                                      43.808975   
11899                                     41.178518   
895                                       35.920319   
...                                             ...   
11284                                     53.179053   
11964                                     46.784600   
5390                                      43.213846   
860                                       49.933749   
15795                                     50.472256   

       Excess kurtosis of the integrated profile  \
14453                                   0.191285   
15797                                   0.083099   
4510                                    0.233088   
11899                                   0.162482   
895                                     0.364792   
...                                          ...   
11284                                  -0.012418   
11964                                   0.218177   
5390                                    0.663456   
860                                    -0.089940   
15795                                   0.346178   

       Skewness of the integrated profile  Mean of the DM-SNR curve  \
14453                            0.007945                  2.921405   
15797                           -0.087226                  1.098662   
4510                             0.451613                  1.142977   
11899                            1.050363                  1.161371   
895                              2.250635                  2.826923   
...                                   ...                       ...   
11284                           -0.556021                  7.186455   
11964                            0.226757                  6.140468   
5390                             0.433088                  0.785117   
860                             -0.226726                  3.859532   
15795                            0.184797                  0.769231   

       Standard deviation of the DM-SNR curve  \
14453                               18.009581   
15797                               11.590431   
4510                                13.774721   
11899                               13.353561   
895                                 17.559245   
...                                       ...   
11284                               29.308266   
11964                               30.271961   
5390                                11.628149   
860                                 21.501505   
15795                               11.792603   

       Excess kurtosis of the DM-SNR curve  Skewness of the DM-SNR curve  
14453                             8.468258                     84.137111  
15797                            14.180137                    242.867752  
4510                             14.148388                    215.404864  
11899                            14.472023                    233.870475  
895                               8.733436                     90.352868  
...                                    ...                           ...  
11284                             4.531382                     21.725143  
11964                             5.732201                     34.357282  
5390                             17.055215                    312.204325  
860                               7.

In [ ]:
from sklearn.svm import SVC

# Initialize SVC with a linear kernel and C=100.0
svc_linear = SVC(kernel='linear', C=1000.0, random_state=42)

# Train the model
svc_linear.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred_linear = svc_linear.predict(X_test_scaled)

# Evaluate the model
accuracy_linear = accuracy_score(y_test, y_pred_linear)
conf_matrix_linear = confusion_matrix(y_test, y_pred_linear)
class_report_linear = classification_report(y_test, y_pred_linear)

print(f"Accuracy (Linear Kernel, C=1000.0): {accuracy_linear:.4f}")
print("\nConfusion Matrix (Linear Kernel, C=1000.0):\n", conf_matrix_linear)
print("\nClassification Report (Linear Kernel, C=1000.0):\n", class_report_linear)

Accuracy (Linear Kernel, C=100.0): 0.9791

Confusion Matrix (Linear Kernel, C=100.0):
 [[3243   16]
 [  59  262]]

Classification Report (Linear Kernel, C=100.0):
               precision    recall  f1-score   support

           0       0.98      1.00      0.99      3259
           1       0.94      0.82      0.87       321

    accuracy                           0.98      3580
   macro avg       0.96      0.91      0.93      3580
weighted avg       0.98      0.98      0.98      3580



In [ ]:
from sklearn.svm import SVC

# Initialize SVC with a linear kernel and C=100.0
svc_linear = SVC(kernel='linear', C=100.0, random_state=42)

# Train the model
svc_linear.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred_linear = svc_linear.predict(X_test_scaled)

# Evaluate the model
accuracy_linear = accuracy_score(y_test, y_pred_linear)
conf_matrix_linear = confusion_matrix(y_test, y_pred_linear)
class_report_linear = classification_report(y_test, y_pred_linear)

print(f"Accuracy (Linear Kernel, C=100.0): {accuracy_linear:.4f}")
print("\nConfusion Matrix (Linear Kernel, C=100.0):\n", conf_matrix_linear)
print("\nClassification Report (Linear Kernel, C=100.0):\n", class_report_linear)

Accuracy (Linear Kernel, C=100.0): 0.9791

Confusion Matrix (Linear Kernel, C=100.0):
 [[3243   16]
 [  59  262]]

Classification Report (Linear Kernel, C=100.0):
               precision    recall  f1-score   support

           0       0.98      1.00      0.99      3259
           1       0.94      0.82      0.87       321

    accuracy                           0.98      3580
   macro avg       0.96      0.91      0.93      3580
weighted avg       0.98      0.98      0.98      3580



In [ ]:
from sklearn.svm import SVC

# Initialize SVC with a linear kernel and C=100.0
svc_linear = SVC(kernel='linear', C=100.0, random_state=42)

# Train the model
svc_linear.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred_linear = svc_linear.predict(X_test_scaled)

# Evaluate the model
accuracy_linear = accuracy_score(y_test, y_pred_linear)
conf_matrix_linear = confusion_matrix(y_test, y_pred_linear)
class_report_linear = classification_report(y_test, y_pred_linear)

print(f"Accuracy (Linear Kernel, C=100.0): {accuracy_linear:.4f}")
print("\nConfusion Matrix (Linear Kernel, C=100.0):\n", conf_matrix_linear)
print("\nClassification Report (Linear Kernel, C=100.0):\n", class_report_linear)

In [ ]:
from sklearn.svm import SVC

# Initialize SVC with a polynomial kernel and C=1.0
svc_poly = SVC(kernel='poly', C=1.0, random_state=42)

# Train the model
svc_poly.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred_poly = svc_poly.predict(X_test_scaled)

# Evaluate the model
accuracy_poly = accuracy_score(y_test, y_pred_poly)
conf_matrix_poly = confusion_matrix(y_test, y_pred_poly)
class_report_poly = classification_report(y_test, y_pred_poly)

print(f"Accuracy (Polynomial Kernel, C=1.0): {accuracy_poly:.4f}")
print("\nConfusion Matrix (Polynomial Kernel, C=1.0):\n", conf_matrix_poly)
print("\nClassification Report (Polynomial Kernel, C=1.0):\n", class_report_poly)

Accuracy (Polynomial Kernel, C=1.0): 0.9785

Confusion Matrix (Polynomial Kernel, C=1.0):
 [[3246   13]
 [  64  257]]

Classification Report (Polynomial Kernel, C=1.0):
               precision    recall  f1-score   support

           0       0.98      1.00      0.99      3259
           1       0.95      0.80      0.87       321

    accuracy                           0.98      3580
   macro avg       0.97      0.90      0.93      3580
weighted avg       0.98      0.98      0.98      3580



In [ ]:
from sklearn.svm import SVC

# Initialize SVC with a linear kernel and C=100.0
svc_linear = SVC(kernel='linear', C=100.0, random_state=42)

# Train the model
svc_linear.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred_linear = svc_linear.predict(X_test_scaled)

# Evaluate the model
accuracy_linear = accuracy_score(y_test, y_pred_linear)
conf_matrix_linear = confusion_matrix(y_test, y_pred_linear)
class_report_linear = classification_report(y_test, y_pred_linear)

print(f"Accuracy (Linear Kernel, C=100.0): {accuracy_linear:.4f}")
print("\nConfusion Matrix (Linear Kernel, C=100.0):\n", conf_matrix_linear)
print("\nClassification Report (Linear Kernel, C=100.0):\n", class_report_linear)

Accuracy (Linear Kernel, C=100.0): 0.9791

Confusion Matrix (Linear Kernel, C=100.0):
 [[3243   16]
 [  59  262]]

Classification Report (Linear Kernel, C=100.0):
               precision    recall  f1-score   support

           0       0.98      1.00      0.99      3259
           1       0.94      0.82      0.87       321

    accuracy                           0.98      3580
   macro avg       0.96      0.91      0.93      3580
weighted avg       0.98      0.98      0.98      3580



In [ ]:
from sklearn.svm import SVC

# Initialize SVC with a sigmoid kernel and C=1.0
svc_sigmoid_c1 = SVC(kernel='sigmoid', C=1.0, random_state=42)

# Train the model
svc_sigmoid_c1.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred_sigmoid_c1 = svc_sigmoid_c1.predict(X_test_scaled)

# Evaluate the model
accuracy_sigmoid_c1 = accuracy_score(y_test, y_pred_sigmoid_c1)
conf_matrix_sigmoid_c1 = confusion_matrix(y_test, y_pred_sigmoid_c1)
class_report_sigmoid_c1 = classification_report(y_test, y_pred_sigmoid_c1)

print(f"Accuracy (Sigmoid Kernel, C=1.0): {accuracy_sigmoid_c1:.4f}")
print("\nConfusion Matrix (Sigmoid Kernel, C=1.0):\n", conf_matrix_sigmoid_c1)
print("\nClassification Report (Sigmoid Kernel, C=1.0):\n", class_report_sigmoid_c1)

Accuracy (Sigmoid Kernel, C=1.0): 0.8751

Confusion Matrix (Sigmoid Kernel, C=1.0):
 [[3034  225]
 [ 222   99]]

Classification Report (Sigmoid Kernel, C=1.0):
               precision    recall  f1-score   support

           0       0.93      0.93      0.93      3259
           1       0.31      0.31      0.31       321

    accuracy                           0.88      3580
   macro avg       0.62      0.62      0.62      3580
weighted avg       0.88      0.88      0.88      3580



In [ ]:
from sklearn.svm import SVC

# Initialize SVC with a sigmoid kernel and C=100.0
svc_sigmoid_c100 = SVC(kernel='sigmoid', C=100.0, random_state=42)

# Train the model
svc_sigmoid_c100.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred_sigmoid_c100 = svc_sigmoid_c100.predict(X_test_scaled)

# Evaluate the model
accuracy_sigmoid_c100 = accuracy_score(y_test, y_pred_sigmoid_c100)
conf_matrix_sigmoid_c100 = confusion_matrix(y_test, y_pred_sigmoid_c100)
class_report_sigmoid_c100 = classification_report(y_test, y_pred_sigmoid_c100)

print(f"Accuracy (Sigmoid Kernel, C=100.0): {accuracy_sigmoid_c100:.4f}")
print("\nConfusion Matrix (Sigmoid Kernel, C=100.0):\n", conf_matrix_sigmoid_c100)
print("\nClassification Report (Sigmoid Kernel, C=100.0):\n", class_report_sigmoid_c100)

Accuracy (Sigmoid Kernel, C=100.0): 0.8749

Confusion Matrix (Sigmoid Kernel, C=100.0):
 [[3033  226]
 [ 222   99]]

Classification Report (Sigmoid Kernel, C=100.0):
               precision    recall  f1-score   support

           0       0.93      0.93      0.93      3259
           1       0.30      0.31      0.31       321

    accuracy                           0.87      3580
   macro avg       0.62      0.62      0.62      3580
weighted avg       0.88      0.87      0.88      3580



In [ ]:
from sklearn.svm import SVC

# Initialize SVC with a sigmoid kernel and C=1000.0
svc_sigmoid_c1000 = SVC(kernel='sigmoid', C=1000.0, random_state=42)

# Train the model
svc_sigmoid_c1000.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred_sigmoid_c1000 = svc_sigmoid_c1000.predict(X_test_scaled)

# Evaluate the model
accuracy_sigmoid_c1000 = accuracy_score(y_test, y_pred_sigmoid_c1000)
conf_matrix_sigmoid_c1000 = confusion_matrix(y_test, y_pred_sigmoid_c1000)
class_report_sigmoid_c1000 = classification_report(y_test, y_pred_sigmoid_c1000)

print(f"Accuracy (Sigmoid Kernel, C=1000.0): {accuracy_sigmoid_c1000:.4f}")
print("\nConfusion Matrix (Sigmoid Kernel, C=1000.0):\n", conf_matrix_sigmoid_c1000)
print("\nClassification Report (Sigmoid Kernel, C=1000.0):\n", class_report_sigmoid_c1000)

Accuracy (Sigmoid Kernel, C=1000.0): 0.8749

Confusion Matrix (Sigmoid Kernel, C=1000.0):
 [[3033  226]
 [ 222   99]]

Classification Report (Sigmoid Kernel, C=1000.0):
               precision    recall  f1-score   support

           0       0.93      0.93      0.93      3259
           1       0.30      0.31      0.31       321

    accuracy                           0.87      3580
   macro avg       0.62      0.62      0.62      3580
weighted avg       0.88      0.87      0.88      3580



based on the accuracy scores from our experiments, the Linear Kernel achieved the best overall accuracy, at approximately 0.9791 (or 97.91%). Both C=1.0 and C=100.0 for the linear kernel yielded this top accuracy. The polynomial kernel was very close at about 0.9785, while the sigmoid kernel performed significantly lower.